x-pack/plugins/security_solution/server/lib/detection_engine/rule_execution_log/event_log_adapter/event_log_client.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IEventLogger, IEventLogService } from '../../../../../../event_log/server';
+import { RuleExecutionStatus } from '../../../../../common/detection_engine/schemas/common/schemas';
+import { LogStatusChangeArgs } from '../types';
+import { RuleExecutionLogAction, RULE_EXECUTION_LOG_PROVIDER } from './constants';
+
+const statusSeverityDict: Record<RuleExecutionStatus, number> = {
+  [RuleExecutionStatus.succeeded]: 0,
+  [RuleExecutionStatus['going to run']]: 10,
+  [RuleExecutionStatus.warning]: 20,
+  [RuleExecutionStatus['partial failure']]: 20,
+  [RuleExecutionStatus.failed]: 30,
+};
+
+interface FindExecutionLogArgs {
+  ruleIds: string[];
+  spaceId: string;
+  logsCount?: number;
+  statuses?: RuleExecutionStatus[];
+}
+
+interface LogExecutionMetricsArgs {
+  ruleId: string;
+  spaceId: string;
+  metrics: EventLogExecutionMetrics;
+}
+
+interface EventLogExecutionMetrics {
+  searchDurationSum?: number;
+  indexingDurationSum?: number;
+  executionGapDuration?: number;
+}
+
+interface IEventLogClient {
+  find: (args: FindExecutionLogArgs) => Promise<{}>;
+  logStatusChange: (args: LogStatusChangeArgs) => Promise<void>;
+  logExecutionMetrics: (args: LogExecutionMetricsArgs) => Promise<void>;
+}
+
+export class EventLogClient implements IEventLogClient {
+  private sequence = 0;
+  private eventLogger: IEventLogger;
+
+  constructor(eventLog: IEventLogService) {
+    this.eventLogger = eventLog.getLogger({
+      event: { provider: RULE_EXECUTION_LOG_PROVIDER },
+    });
+  }
+
+  public async find({ ruleIds, spaceId, statuses, logsCount = 1 }: FindExecutionLogArgs) {
+    return {}; // TODO implement
+  }
+
+  public async logExecutionMetrics({
+    ruleId,
+    metrics,
+    spaceId, // TODO write spaceId
+  }: LogExecutionMetricsArgs) {
+    this.eventLogger.logEvent({
```

Comment:
I'm including below a complete event log doc that we are writing for rules - in this case, the event written after a rule executes.

<details>
<summary>alerting event log doc</summary>

```json
{
  "@timestamp" : "2021-07-27T02:27:30.156Z",
  "event" : {
    "provider" : "alerting",
    "action" : "execute",
    "kind" : "alert",
    "category" : [
      "stackAlerts"
    ],
    "start" : "2021-07-27T02:27:30.156Z",
    "outcome" : "success",
    "end" : "2021-07-27T02:27:30.237Z",
    "duration" : 81000000
  },
  "kibana" : {
    "saved_objects" : [
      {
        "rel" : "primary",
        "type" : "alert",
        "id" : "f2683000-ebbe-11eb-b20b-d74b5b35b7e7",
        "type_id" : ".index-threshold"
      }
    ],
    "task" : {
      "scheduled" : "2021-07-27T02:27:25.164Z",
      "schedule_delay" : 4992000000
    },
    "alerting" : {
      "status" : "ok"
    },
    "server_uuid" : "e520efe6-f6c4-41d2-8a7e-1ec90211da3a"
  },
  "rule" : {
    "id" : "f2683000-ebbe-11eb-b20b-d74b5b35b7e7",
    "license" : "basic",
    "category" : ".index-threshold",
    "ruleset" : "stackAlerts",
    "name" : "stressing index-threshold 3/20"
  },
  "message" : "alert executed: .index-threshold:f2683000-ebbe-11eb-b20b-d74b5b35b7e7: 'stressing index-threshold 3/20'",
  "ecs" : {
    "version" : "1.8.0"
  }
}
```

</details>

It would be great to fill out as many of these as you can, but it can be difficult because we don't always have all the field values at the point where we write the logs.  Filling some of these in would allow us to see some of the same docs across searches, but they are also occasionally **INVALUABLE** during problem diagnosis.

Here are the most interesting ones:
- `rule.name` - name of the rule, which is super-useful, but often not available where the events are logged :-(
- `rule.category` - rule type id
- `kibana.saved_objects` - this is what we use for our "RBAC by passing in SO ids which we validate you can read; you'd need this if you want to use that kind of RBAC till we have the real one.  I don't think the `type_id` is required here - I think we added it before starting to populate `rule.category`
- `kibana.server_uuid` - we write this to identify the specific Kibana instance when there are more than one
- `rule.ruleset` - producer - probably not immediately or maybe even long-term useful, not sure

We actively use `rule.name` and `rule.category` for various queries / graphs over rules and their types.  And `rule.name` is so much nicer than having to deal with just the UUID id.

It's also possible you may want to have a common date if you are writing multiple documents - or use some other date that perhaps got passed along, for use in correlating events, rather than `new Date().toISOString()` for `@timestamp`.  Otherwise, if you just want the "current" date, you don't need to provide the field at all, as the event log will add one with the same expression.

Reply:
Yeah, we've overlooked these fields, especially `kibana.saved_objects` is an important one to fill, otherwise we wouldn't be able to find them. Thank you for catching this 👍

Reply:
Thank you, @pmuellr, for your suggestion 👍 Added `rule.name`, `rule.category`, and `kibana.saved_objects` and dropped `@timestamp`. I'm not sure about `rule.ruleset`, so decided not to add it right now.